tiny_dnn/core/kernels/global_avepool_op_avx.h

```diff
@@ -0,0 +1,211 @@
+/*
+    Copyright (c) 2017, Taiga Nomi
+    All rights reserved.
+
+    Use of this source code is governed by a BSD-style license that can be found
+    in the LICENSE file.
+*/
+#pragma once
+
+namespace tiny_dnn {
+namespace kernels {
+
+#ifdef CNN_USE_AVX
+#ifdef CNN_USE_DOUBLE
+
+// double version
+inline void global_avepool_op_avx(const tensor_t &in_data,
+                                  tensor_t &out_data,
+                                  const core::global_avepool_params &params,
+                                  const bool layer_parallelize) {
+  const size_t pool_area            = params.in.width_ * params.in.height_;
+  const double pool_area_inv        = 1.0 / pool_area;
+  const size_t nblocks_per_channel  = pool_area / 4;
+  const size_t nremains_per_channel = pool_area & 3;
+
+  int64_t mask_src[] = {-1, -1, -1, -1, 0, 0, 0, 0};
+  __m256i imask =
+    _mm256_loadu_si256((__m256i const *)(mask_src + 4 - nremains_per_channel));
+
+  for_i(layer_parallelize, in_data.size(), [&](size_t sample) {
+    const vec_t &in = in_data[sample];
+    vec_t &out      = out_data[sample];
+
+    for (size_t i = 0; i < params.in.depth_; i++) {
+      __m256d sum_m            = _mm256_setzero_pd();
+      const size_t depth_index = i * pool_area;
+
+      // unrolling the loop 8x and 4x times for maximum throughput
+      const size_t loop_8x = nblocks_per_channel / 8;
+      const size_t loop_4x = (nblocks_per_channel - loop_8x * 8) / 4;
+      for (size_t j = 0; j < 8 * loop_8x; j += 8) {
+        __m256 in_m0 = _mm256_loadu_pd(&in[depth_index + 4 * j]);
```

Comment:
For double version of `global_avepool_op_avx`, you need to change type of these local variables from `__m256` to `__m256d`.